handler/src/main/java/io/netty/handler/ssl/TlsMasterKeyHandler.java

```diff
@@ -0,0 +1,166 @@
+
+    /**
+     * The JRE SSLSessionImpl cannot be imported
+     */
+    private static Class<?> SSL_SESSIONIMPL_CLASS;
+
+    /**
+     * The master key field in the SSLSessionImpl
+     */
+    private static Field SSL_SESSIONIMPL_MASTER_SECRET_FIELD;
+
+    /**
+     * A system property that can be used to turn on/off the {@link TlsMasterKeyHandler} dynamically without having
+     * to edit your pipeline.
+     * <code>-Dio.netty.tls.masterKeyHandler=true</code>
+     */
+    private static final String SYSTEM_PROP_KEY = "io.netty.tls.masterKeyHandler";
+
+    /**
+     * The unavailability clause of whether the private Sun implementation of SSLSessionImpl is available.
+     */
+    private static final Throwable UNAVAILABILITY_CAUSE;
+
+    static {
+        Throwable cause = null;
+        try {
+            SSL_SESSIONIMPL_CLASS = Class.forName("sun.security.ssl.SSLSessionImpl");
+            SSL_SESSIONIMPL_MASTER_SECRET_FIELD = SSL_SESSIONIMPL_CLASS.getDeclaredField("masterSecret");
+            SSL_SESSIONIMPL_MASTER_SECRET_FIELD.setAccessible(true);
+        } catch (Throwable e) {
+            cause = e;
+            logger.debug("sun.security.ssl.SSLSessionImpl is unavailable.", e);
+        }
+        UNAVAILABILITY_CAUSE = cause;
+    }
+
+    /**
+     * Constructor.
+    */
+    public TlsMasterKeyHandler() {
+    }
+
+    /**
+     * Ensure that SSLSessionImpl is available.
+     * @throws UnsatisfiedLinkError if unavailable
+     */
+    public static void ensureSunSslEngineAvailability() {
+        if (UNAVAILABILITY_CAUSE != null) {
+            throw (Error) new IllegalStateException(
+                    "Failed to find SSLSessionImpl on classpath").initCause(UNAVAILABILITY_CAUSE);
+        }
+    }
+
+    /**
+     * Returns the cause of unavailability.
+     *
+     * @return the cause if unavailable. {@code null} if available.
+     */
+    public static Throwable sunSslEngineUnavailabilityCause() {
+        return UNAVAILABILITY_CAUSE;
+    }
+
+    /* Returns {@code true} if and only if sun.security.ssl.SSLSessionImpl exists in the runtime.
+     */
+    public static boolean isSunSslEngineAvailable() {
+        return UNAVAILABILITY_CAUSE == null;
+    }
+
+    /**
+     * Consume the master key for the session and the sessionId
+     * @param masterKey A 48-byte secret shared between the client and server.
+     * @param sessionId The sessionId An arbitrary byte sequence chosen by the server to identify an
+     *                  active or resumable session state.
+     */
+    protected abstract void accept(SecretKey masterKey, byte[] sessionId);
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+        //only try to log the session info if the ssl handshake has successfully completed.
+        if (evt == SslHandshakeCompletionEvent.SUCCESS) {
+
+            boolean shouldHandle = SystemPropertyUtil.getBoolean(SYSTEM_PROP_KEY, false);
```

Comment:
should re really lookup the property every time ?

Reply:
I've left this in for now -- I'm open to making it once lookup.
I imagined this is largely a test driven handler anyways not to be included in typical pipelines?